Add support for elements validation in DataForm's array fields

elazzabi opened the pull request:
## What?

This adds support for validation in DataForm's array fields. The following prop can be used to only accept input from the elements:

```
isValid: {
	elements: true,
}
```

## Why?

This allows the component to be customized depending on use cases.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions

Add this to any Storybook implementation to test this.

```
isValid: {
	elements: true,
}
```

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

This is the only mode currently supported:

https://github.com/user-attachments/assets/e22e53c7-f629-4e90-a4ad-af8b1b7bb4f3

This is the new mode:

https://github.com/user-attachments/assets/4d113200-d385-4798-a301-fffb8b98ae71

elazzabi committed:
Add support for validating elements

elazzabi committed:
Add validation input and how to

elazzabi committed:
Add array validation with required param

elazzabi committed:
Update validation logic to check if the token matches any available elements using `findElementByLabel` instead of `isItemValid`.

elazzabi committed:
Enhance validation logic to support array types by ensuring all values are valid elements and streamline single-value checks for element validity.

elazzabi committed:
Add tests for array and text field validations and validating against specified elements.

oandregal reviewed (commented):
Nice progress here. I think the missing bit is how/whether the `FormTokenField` can display messages to users (validation errors) and/or signal invalid state via color, like [other controls do](https://wordpress.github.io/gutenberg/?path=/story/dataviews-dataform--validation). cc @mirka @jameskoster for thoughts

oandregal commented on packages/dataviews/src/validation.ts:
```diff
@@ -44,6 +54,26 @@ export function isItemValid< Item >(
+					return false;
+				}
+				// For single-value fields, check if the value is a valid element
+				return validValues.includes( value );
```
In addition to this, we need to display what's happening to the user.

Take the [text control](https://github.com/WordPress/gutenberg/blob/b7b321ab41f81e41bf46c99e035265b04f8f9785/packages/dataviews/src/dataform-controls/text.tsx#L36) as an example. We handle `isValid.required` and `isValid.custom` there. By introducing `isValid.elements` we need to communicate that as well (we could use the `customValidator` function of the control to display something along the lines of "Value must be one of the elements".

elazzabi replied:
I added an error when this happens
<img width="638" height="284" alt="CleanShot 2025-09-11 at 09 30 27@2x" src="https://github.com/user-attachments/assets/3e1c7a36-76b6-4278-8343-c2993e999183" />

elazzabi committed:
Add categories field to DataFormValidationComponent with validation rules and predefined options

mirka commented:
> how/whether the `FormTokenField` can display messages to users (validation errors) and/or signal invalid state via color, like [other controls do](https://wordpress.github.io/gutenberg/?path=/story/dataviews-dataform--validation).

Yes, I think there can be a `ValidatedFormTokenControl`.

elazzabi committed:
Add ValidatedFormTokenControl component to enhance array input validation with custom and required checks

elazzabi committed:
Enhance array validation logic to ensure elements are only checked if isValid elements is passed

elazzabi committed:
Refactor array control to support internationalization and improve value handling.

elazzabi committed:
Add countries field to DataFormValidationComponent with validation rules and predefined options

elazzabi commented:
@mirka @oandregal I added a `ValidatedFormTokenControl` to wrap the component with some messaging. I also added 2 validation examples to the Storybook. Here's the current behavior:

https://github.com/user-attachments/assets/835b4ab7-c9f5-4a19-b0f2-e86a7a21e684

oandregal commented on packages/dataviews/src/validation.ts:
```diff
@@ -34,6 +42,8 @@ export function isItemValid< Item >(
 				( field.type === 'integer' &&
 					isEmptyNullOrUndefined( value ) ) ||
+				( field.type === 'array' &&
+					isNotArrayEmptyOrContainsInvalidElements( value ) ) ||
```
This is a small nit. For a second, I was confused for the use of "invalid elements" here, I thought _why are we validating isValid.elements here if there's a specific check for it_? Perhaps something along these lines makes that more clear?

```suggestion
					isArrayOrElementsEmptyNullOrUndefined( value ) ) ||
```

oandregal commented on packages/dataviews/src/field-types/array.tsx:
```diff
@@ -59,7 +59,7 @@ const arrayFieldType: FieldTypeDefinition< any > = {
 			}
 
-			if ( field?.elements ) {
+			if ( field?.elements && field.isValid?.elements ) {
```
Given `isValid.elements` is implemented in validation.ts, do we still need it here? (same for any other field type)

elazzabi replied:
I added this because, without it, the validation is run whenever the `elements` are present. Including when it is okay to add new elements 

https://github.com/user-attachments/assets/be3a317a-c82a-48fb-94b1-b60b3301b27c

oandregal replied:
Is this code ever executed? I'd think we can remove it because it's covered by the control [here](https://github.com/WordPress/gutenberg/blob/661758485a0dc5c2d9dedae50dc92124a7a4720e/packages/dataviews/src/dataform-controls/array.tsx#L69), and so it never checks the custom validation [here](https://github.com/WordPress/gutenberg/blob/661758485a0dc5c2d9dedae50dc92124a7a4720e/packages/dataviews/src/dataform-controls/array.tsx#L77) (that will run the same check again for elements).

There are a few levels to validation:

- Control-level: the array control handles element validation, providing the corresponding message if it's invalid.
- Form-level, via the `isItemValid` utility: this handles element validation (this is for form-level).

By introducing `isValid.elements` I understand that we can:

- remove the existing checks in all field types (previously bundled as part of `isValid.custom`)
- make sure every control handles `isValid.elements` properly (done for the array control)
- make sure `isItemValid` also handles it ([done](https://github.com/WordPress/gutenberg/blob/661758485a0dc5c2d9dedae50dc92124a7a4720e/packages/dataviews/src/validation.ts#L62))

Does this make sense, or am I missing any context here?

elazzabi replied:
After seeing it closely, I agree that this part was inadvertently executed (as I recorded in the video), but it should be removed altogether f2e301d40ff34bd97cacdef411de311eb4dddaf6

Now that it is removed, and regarding this part:

> make sure every control handles isValid.elements properly (done for the array control)

I only see the `select` control that might benefit from such validation. Should I implement it there, too?

oandregal reviewed (commented):
I ran into this (adding two elements, the first in `Titlecase` the second in `lowercase`):

https://github.com/user-attachments/assets/9d4c6566-882a-4133-bd7f-3033035421bc


It doesn't happen in the FormToken component:

https://github.com/user-attachments/assets/22c7e180-1119-4453-9ee5-a0b0cf71cad9

mirka commented on packages/dataviews/src/dataform-controls/array.tsx:
```diff
@@ -1,14 +1,94 @@
  */
 import type { DataFormControlProps } from '../types';
 
+// Create a validated FormTokenField wrapper
+const ValidatedFormTokenControl = ( {
```
Sorry for being unclear, I meant we can add a `ValidatedFormTokenControl` component to `@wordpress/components`, along with the [other ones](https://wordpress.github.io/gutenberg/?path=/docs/components-validated-form-controls-overview--docs) 😄

Please open a separate PR for this addition so we can do a component review. And do let me know if you find anything lacking in the existing documentation 🙏

elazzabi replied:
Oh okay. @oandregal I'll implement this change ^ first before continuing, hopefully it will reduce the diff here 😄

elazzabi replied:
@mirka I created a PR for the component here https://github.com/WordPress/gutenberg/pull/71350

Let me know if you have any feedback

elazzabi committed:
Merge branch 'trunk' into add/elements-validation

elazzabi committed:
Expose ValidatedFormTokenField from private API

elazzabi committed:
Remove customValidation

elazzabi committed:
Import ValidatedFormTokenField and use it in Array

elazzabi committed:
Add displayTransform and __experimentalRenderItem

elazzabi committed:
Replace arrays for display const and remove unused findElementByValue

elazzabi committed:
Adapt suggestions to pass values

elazzabi committed:
Add customValidity and onFocus

elazzabi committed:
Add validation with validateTokens

elazzabi commented on packages/dataviews/src/dataform-controls/array.tsx:
```diff
@@ -18,68 +22,180 @@ export default function ArrayControl< Item >( {
-			__next40pxDefaultSize
-			__nextHasNoMarginBottom
+			__experimentalShowHowTo={ ! field.isValid?.elements }
+			displayTransform={ ( token: any ) => {
```
I decided to use the `displayTransform` instead of using the label to token conversion, given that it's fragile.

elazzabi committed:
Rename function

elazzabi committed:
Add changelogs

oandregal commented on packages/components/CHANGELOG.md:
```diff
@@ -11,6 +11,8 @@
 -   `ValidatedCheckboxControl`: Expose the component under private API's ([#71505](https://github.com/WordPress/gutenberg/pull/71505/)).
 -   Expose `ValidatedTextareaControl` via Private APIs ([#71495](https://github.com/WordPress/gutenberg/pull/71495))
 -   Add support for ValidatedFormTokenField [#71350](https://github.com/WordPress/gutenberg/pull/71350).
+-   Add support for elements validation in DataForm's array fields [#71194](https://github.com/WordPress/gutenberg/pull/71194)
```
```suggestion
-   Expose `ValidatedFormTokenField` via private APIs [#71194](https://github.com/WordPress/gutenberg/pull/71194)
```

this is the component's changelog :)

elazzabi committed:
Change changelog for components

Co-authored-by: André <583546+oandregal@users.noreply.github.com>

elazzabi committed:
Call onInputChange after reseting the input

elazzabi committed:
Pass an onInputChange instead of onFocus

elazzabi committed:
Merge branch 'add/elements-validation' of https://github.com/elazzabi/gutenberg into add/elements-validation

elazzabi committed:
Remove validation for elements as it is replaced by control-level validation

elazzabi committed:
Add documentation for elements

oandregal commented on packages/components/src/form-token-field/index.tsx:
```diff
@@ -185,6 +185,7 @@ export function FormTokenField( props: FormTokenFieldProps ) {
 		} else {
 			// Reset to initial state
 			setIncompleteTokenValue( '' );
+			onInputChange( '' );
```
Could you clarify what is this for?

I'm not super familiar with the FormToken component, but this is not clearing internal state. It's triggering the onInputChange user callback when the component loses focus under certain conditions. It seems weird.

oandregal replied:
Tried testing the components themselves (both validated and the normal version) but I don't know how to trigger this path. It'd be great if you could share more about this.

elazzabi replied:
Sure!

The `FormToken` component accepts a `onInputChange` prop that I think should be called when the input is changed. Here, when the user clicks outside the input, we `setIncompleteTokenValue( '' )` but we were not letting the parent component know about the change.

In practice, this is how it looks without `onInputChange`:

https://github.com/user-attachments/assets/8d1e7ee8-0f4f-4343-8493-54010ff5268c

This is how it looks with it:

https://github.com/user-attachments/assets/ea38949a-e5ce-4347-ba10-91ada9f662f1

(The error message is cleaned after navigating away)

In summary, we need to inform the parent about the input change to remove the error message.

oandregal replied:
I continued the conversation here https://github.com/WordPress/gutenberg/pull/71194#discussion_r2363039634

oandregal commented on packages/dataviews/src/dataform-controls/array.tsx:
```diff
@@ -18,68 +22,183 @@ export default function ArrayControl< Item >( {
+						...data,
+						[ id ]: tokenValues,
+					},
+					field
```
When rebasing, this needs to be something along these lines ([example](https://github.com/WordPress/gutenberg/blob/trunk/packages/dataviews/src/dataform-controls/checkbox.tsx#L42)). See [PR](https://github.com/WordPress/gutenberg/pull/71604), [docs](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-dataviews/#getvalue-and-setvalue) and the [story](https://wordpress.github.io/gutenberg/?path=/story/dataviews-dataform--data-adapter) for context:

```js
const result = field.isValid?.custom?.(
  deepMerge(
      data,
      setValue( { item: data, value: newValue, } ) as Partial< Item >
  ),
  field
);
```

oandregal commented on packages/dataviews/src/dataform-controls/array.tsx:
```diff
@@ -18,68 +22,183 @@ export default function ArrayControl< Item >( {
 
 			onChange( {
-				[ id ]: stringTokens,
+				[ id ]: valueTokens,
```
When rebasing, this needs to be something along these lines ([example](https://github.com/WordPress/gutenberg/blob/trunk/packages/dataviews/src/dataform-controls/checkbox.tsx#L34)). See [PR](https://github.com/WordPress/gutenberg/pull/71604), [docs](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-dataviews/#getvalue-and-setvalue) and the [story](https://wordpress.github.io/gutenberg/?path=/story/dataviews-dataform--data-adapter) for context:

```js
onChange(
  setValue( { item: data, value: valueTokens } )
);
```

oandregal commented on packages/dataviews/src/dataform-controls/array.tsx:
```diff
@@ -18,68 +22,183 @@ export default function ArrayControl< Item >( {
-				elements?.map( ( suggestion ) => suggestion.label ) ?? []
-			}
+			suggestions={ elements?.map( ( element ) => element.value ) }
+			messages={ {
+				added: __( 'Item added.' ),
+				removed: __( 'Item removed.' ),
+				remove: __( 'Remove item' ),
+				__experimentalInvalid: customInvalidMessage,
+			} }
```
The added/removed/remove messages are the same as the ones provided by the control by default. The __experimentalInvalid is the same as the one provided by the `customValidity`. So I wonder if this is ever used?

This is what I did to test:

- Set value in this component to `value={ [ { value: 'howdy', label: 'Howdy' } ] }` so there's an invalid value not part of the elements.
- Changed the `customInvalidMessage` to `MESSAGE`.
- Tested the story, but I was only able to get the `customValidity` message:

<img width="385" height="315" alt="Screenshot 2025-09-17 at 11 41 11" src="https://github.com/user-attachments/assets/177e2b40-4a6a-4b36-ad22-5ef96ba2619c" />

elazzabi replied:
After testing this, I agree. That wasn't used, and having `onValidate` and the `__experimental` flag is all we need.

a9872b29ef123979bc1e6db13f0bd06bf97343c1

oandregal commented on packages/dataviews/src/dataform-controls/array.tsx:
```diff
@@ -18,68 +22,183 @@ export default function ArrayControl< Item >( {
+				[ id ]: valueTokens,
 			} );
+
+			validateTokens( tokens );
```
`onValidate` is called every time `onChange` is called, right? So this would trigger validation twice.

elazzabi replied:
Great catch! Indeed, it was running twice. I don't think we need it here, so I removed it 08f9b87d26c7d795b4603602724d17a2dcf8839b

oandregal commented on packages/dataviews/src/dataform-controls/array.tsx:
```diff
@@ -18,68 +22,183 @@ export default function ArrayControl< Item >( {
-			value={ arrayValue }
+			value={ arrayValueAsElements }
 			onChange={ onChangeControl }
+			onInputChange={ onInputChange }
```
So, `onValidate` triggers for every `onChange` event. And the user is never able to introduce invalid data as per `__experimentalValidateInput`. So, the only case in which we ever display an error for `isValid.elements` is when the case in which the component loads with bad data (as [here](https://github.com/WordPress/gutenberg/pull/71194/files#r2354928714)). Is this correct?

If so, why we need this? The custom error for elements will never happen as a result of an input event, so it doesn't need to be cleared when input is empty. Perhaps I'm not considering another use case?

elazzabi replied:
The only reason for this is to catch this edge case, where the input is cleared but the message stays:

https://github.com/user-attachments/assets/25c63133-9f16-4fe6-8330-8f1b16b6d796

oandregal replied:
@elazzabi I've pushed https://github.com/WordPress/gutenberg/pull/71194/commits/20410e48e6e9995bf14cd1982f6361599b984c72 to illustrate my point. It essentially removes everything related to input. We use `__experimentalValidateInput` to prevent an invalid element from being committed, but we don't trigger any validation on input.

I need to look at other things now, but there's this question we need to answer: does it makes sense to just remove `onValidate` / `customValidity` logic altogether? I think the only case where this may be useful is when the component is loaded with data that is not part of the element list.

oandregal replied:
> I think the only case where this may be useful is when the component is loaded with data that is not part of the element list.

So I tested this scenario now by adding `es` to the [data](https://github.com/elazzabi/gutenberg/blob/20410e48e6e9995bf14cd1982f6361599b984c72/packages/dataviews/src/components/dataform/stories/index.story.tsx#L547) that gets loaded. This is the result:

https://github.com/user-attachments/assets/2699ca3e-0df3-4cf8-9776-b595e4813e47

I think we can keep that logic, but the message needs to be more specific. Along the lines of: `Please select from the available options: "es" is invalid.`.

elazzabi replied:
Thanks for the commit @oandregal !

In this case, we are completely removing the feedback when the user tries to insert an element via "," (ie nothing happens), but I'm okay keeping it that way.

I pushed a change to update the message with two versions (singular and plural):
<img width="798" height="234" alt="CleanShot 2025-09-19 at 17 47 41@2x" src="https://github.com/user-attachments/assets/f6b59477-9206-4f79-a24b-bf21558cb6bb" />
<img width="1030" height="208" alt="CleanShot 2025-09-19 at 17 48 07@2x" src="https://github.com/user-attachments/assets/2dbe6aab-124e-4b34-ae9c-f1d4c61a791b" />

elazzabi committed:
Merge branch 'trunk' into add/elements-validation

elazzabi committed:
Add changes to use the new getValue/setValue API

elazzabi committed:
Remove validateToken from onChange

elazzabi committed:
Remove customInvalidMessage and messages

elazzabi committed:
Fix linting

elazzabi committed:
Moves changelog entries to Unreleased

oandregal committed:
Remove input logic

elazzabi committed:
Update message by including tokens

elazzabi committed:
Merge branch 'trunk' into add/elements-validation

elazzabi committed:
Change order of changelog to reflect the timeline
